projects/angular/src/data/datagrid/datagrid-virtual-scroll.directive.spec.ts

```diff
@@ -0,0 +1,229 @@
+    return row.index;
+  }
+
+  setExpanded($event: boolean, item: Row) {
+    item.expanded = $event;
+  }
+}
+
+export default function (): void {
+  describe('ClrDatagrid virtual scroller', function () {
+    function sleep(millisecondsToWait = 100) {
+      return new Promise(resolve => setTimeout(resolve, millisecondsToWait));
+    }
+
+    describe('Typescript API', function () {
+      let fixture: ComponentFixture<any>;
+      let compiled: any;
+      let instance: any;
+
+      beforeEach(async function () {
+        await TestBed.configureTestingModule({
+          imports: [ClarityModule, NoopAnimationsModule],
+          declarations: [FullTest, CustomClrVirtualRowsDirective],
+          schemas: [CUSTOM_ELEMENTS_SCHEMA],
+          providers: DATAGRID_SPEC_PROVIDERS,
+        }).compileComponents();
+
+        fixture = TestBed.createComponent(FullTest);
+
+        compiled = fixture.nativeElement;
+        instance = fixture.componentInstance;
+      });
+
+      afterEach(() => {
+        fixture.destroy();
+      });
+
+      it('allows to manually force a refresh of displayed items when data mutates', function () {
+        fixture.detectChanges();
+
+        expect(instance.virtualScroll.items.all.length).toBe(1000);
+
+        expect(instance.virtualScroll.maxBufferPx).toBe(400);
+        instance.virtualScroll.maxBufferPx = 600;
+        expect(instance.virtualScroll.maxBufferPx).toBe(600);
+
+        expect(instance.virtualScroll.minBufferPx).toBe(200);
+        instance.virtualScroll.minBufferPx = 300;
+        expect(instance.virtualScroll.minBufferPx).toBe(300);
+
+        expect(instance.virtualScroll.itemSize).toBe(24);
+        instance.virtualScroll.itemSize = 20;
+        expect(instance.virtualScroll.itemSize).toBe(20);
+
+        expect(instance.virtualScroll.cdkVirtualForTemplateCacheSize).toBe(4000);
+        instance.virtualScroll.cdkVirtualForTemplateCacheSize = 5000;
+        expect(instance.virtualScroll.cdkVirtualForTemplateCacheSize).toBe(5000);
+      });
+
+      it('Moves focus on PageDown and PageUp', async () => {
+        fixture.autoDetectChanges();
+
+        await fixture.whenStable();
+        await fixture.whenRenderingDone();
+
+        const grid = compiled.querySelector('[role=grid]');
+
+        // need to start with this cell exactly, because it has tabindex=0
+        const headerCheckboxCell = grid.querySelector('[role=columnheader].datagrid-select');
+        headerCheckboxCell.focus();
+        expect(document.activeElement).toEqual(headerCheckboxCell);
+
+        grid.dispatchEvent(new KeyboardEvent('keydown', { code: 'PageDown' }));
+        // active checkbox input with ID clr-dg-row-cb364
+        expect(document.activeElement).toEqual(grid.querySelectorAll('[type=checkbox]')[22]);
+
+        grid.dispatchEvent(new KeyboardEvent('keydown', { code: 'PageDown' }));
+        await sleep();
+        await fixture.whenStable();
+        await sleep();
```

Comment:
One sleep with additional time is enough.
```suggestion
        await sleep(50);
        await fixture.whenStable();
```

Reply:
It's different in latest commits.